lib/init.lua

```diff
@@ -584,57 +598,68 @@ function World.delete(world: World, entityId: i53)
 end
 
 function World.observer(world: World, ...)
-	local componentIds = { ... }
-	
+	local componentIds = {...}
+	local hooks = world.hooks
+
 	return {
-		event = function(event) 
-			local hook = world.hooks[event]
-			world.hooks[event] = nil
+		event = function(event)
+			local hook = hooks[event]
+			hooks[event] = nil
 
 			local last, change
-			return function() 
+			return function()
 				last, change = next(hook, last)
-				if not last then 
+				if not last then
 					return
 				end
 
 				local matched = false
-				
-				while not matched do 
+				local ids = change.ids
+
+				while not matched do
 					local skip = false
-					for _, id in change.ids do 
-						if not table.find(componentIds, id) then 
+					for _, id in ids do
+						if not table.find(componentIds, id) then
 							skip = true
 							break
 						end
 					end
-					
-					if skip then 
+
+					if skip then
 						last, change = next(hook, last)
+						ids = change.ids
 						continue
 					end
 
 					matched = true
 				end
-				
+
 				local queryOutput = {}
+				local length = 0
+
 				local row = change.offset
 				local archetype = change.archetype
 				local columns = archetype.columns
 				local archetypeRecords = archetype.records
-				for _, id in componentIds do 
-					table.insert(queryOutput, columns[archetypeRecords[id]][row])
+				for _, id in componentIds do
```

Comment:
It should never be possible for it not have a value since components are guaranteed to exist under the same archetype.

Reply:
okay good i was just unsure